mopidy/core/library.py

```diff
@@ -69,16 +68,30 @@ def browse(self, uri):
 
         .. versionadded:: 0.18
         """
-        if uri is None:
-            backends = self.backends.with_library_browse.values()
-            unique_dirs = {b.library.root_directory.get() for b in backends}
-            return sorted(unique_dirs, key=operator.attrgetter('name'))
+        return self._roots() if uri is None else self._browse(uri)
 
+    def _roots(self):
+        directories = set()
+        backends = self.backends.with_library_browse.values()
+        futures = {b: b.library.root_directory for b in backends}
+        for backend, future in futures.items():
+            try:
+                directories.add(future.get())
+            except Exception:
+                logger.exception('%s backend caused an exception.',
+                                 backend.actor_ref.actor_class.__name__)
+        return sorted(directories, key=operator.attrgetter('name'))
+
+    def _browse(self, uri):
         scheme = urlparse.urlparse(uri).scheme
         backend = self.backends.with_library_browse.get(scheme)
-        if not backend:
-            return []
-        return backend.library.browse(uri).get()
+        try:
+            if backend:
+                return backend.library.browse(uri).get()  # TODO: sort?
```

Comment:
I don't think its up to us to mess with the order of the results. How are you supposed to make e.g. top lists if we sort the data you return? Remove the comment.